compile error with clang-9

Tried to compile with clang-9 and got the following errors:


```
immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1092:28: error: chosen constructor is explicit in copy-initialization
                    return { pos.shift(), newn, ts, tail };
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1120:16: error: chosen constructor is explicit in copy-initialization
        return { 0, nullptr, new_tail_size, new_tail };
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:201:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_leaf<immer::detail::rbts::full_leaf_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_leaf(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1092:28: error: chosen constructor is explicit in copy-initialization
                    return { pos.shift(), newn, ts, tail };
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:1288:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::full_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1120:16: error: chosen constructor is explicit in copy-initialization
        return { 0, nullptr, new_tail_size, new_tail };
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:108:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_leaf<immer::detail::rbts::leaf_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_leaf(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1092:28: error: chosen constructor is explicit in copy-initialization
                    return { pos.shift(), newn, ts, tail };
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:297:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1120:16: error: chosen constructor is explicit in copy-initialization
        return { 0, nullptr, new_tail_size, new_tail };
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:201:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, false>::visit_leaf<immer::detail::rbts::full_leaf_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_leaf(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1092:28: error: chosen constructor is explicit in copy-initialization
                    return { pos.shift(), newn, ts, tail };
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:1288:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, false>::visit_regular<immer::detail::rbts::full_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1120:16: error: chosen constructor is explicit in copy-initialization
        return { 0, nullptr, new_tail_size, new_tail };
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:108:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, false>::visit_leaf<immer::detail::rbts::leaf_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_leaf(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^
In file included from /home/username/tmp/immer/example/vector/vector.cpp:9:
In file included from /home/username/tmp/immer/immer/vector.hpp:11:
In file included from /home/username/tmp/immer/immer/detail/rbts/rbtree.hpp:13:
/home/username/tmp/immer/immer/detail/rbts/operations.hpp:1092:28: error: chosen constructor is explicit in copy-initialization
                    return { pos.shift(), newn, ts, tail };
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/username/tmp/immer/immer/detail/rbts/position.hpp:297:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, false>::visit_regular<immer::detail::rbts::regular_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/detail/rbts/position.hpp:945:25: note: in instantiation of function template specialization 'immer::detail::rbts::slice_right_visitor<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>, true>::visit_regular<immer::detail::rbts::regular_sub_pos<immer::detail::rbts::node<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6> > &>' requested here
        return Visitor::visit_regular(*this, std::forward<Args>(args)...);
                        ^
/home/username/tmp/immer/immer/vector.hpp:293:20: note: in instantiation of member function 'immer::detail::rbts::rbtree<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
    { return impl_.take(elems); }
                   ^
/home/username/tmp/immer/example/vector/vector.cpp:49:22: note: in instantiation of member function 'immer::vector<int, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::no_transience_policy, false, true>, 5, 6>::take' requested here
        auto v2 = v1.take(3);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:479:19: note: explicit constructor declared here
        constexpr tuple(_UElements&&... __elements)
                  ^


```

--- a/immer/detail/rbts/operations.hpp
+++ b/immer/detail/rbts/operations.hpp
@@ -49,7 +49,7 @@ struct region_for_visitor : visitor_base<region_for_visitor<T>>
 
     template <typename PosT>
     static result_t visit_leaf(PosT&& pos, size_t idx)
-    { return { pos.node()->leaf(), pos.index(idx), pos.count() }; }
+    { return std::make_tuple( pos.node()->leaf(), pos.index(idx), pos.count() ); }
 };
 
 template <typename T>
@@ -896,7 +896,7 @@ struct slice_right_mut_visitor
                         node->inner()[idx] = next;
                         nodr->d.sizes[idx] = last + 1 - ts;
                         nodr->d.count = idx + 1;
-                        return { pos.shift(), node, ts, tail };
+                        return std::make_tuple( pos.shift(), node, ts, tail );
                     } else {
                         auto newn = node_t::copy_inner_r_e(e, node, idx);
                         auto newr = newn->relaxed();
@@ -904,25 +904,25 @@ struct slice_right_mut_visitor
                         newr->d.sizes[idx] = last + 1 - ts;
                         newr->d.count = idx + 1;
                         if (Mutating) pos.visit(dec_visitor{});
-                        return { pos.shift(), newn, ts, tail };
+                        return std::make_tuple( pos.shift(), newn, ts, tail );
                     }
                 } else if (idx == 0) {
                     if (Mutating) pos.visit(dec_right_visitor{}, count_t{1});
-                    return { pos.shift(), nullptr, ts, tail };
+                    return std::make_tuple( pos.shift(), nullptr, ts, tail );
                 } else if (Collapse && idx == 1 && pos.shift() > BL) {
                     auto newn = pos.node()->inner()[0];
                     if (!mutate) newn->inc();
                     if (Mutating) pos.visit(dec_right_visitor{}, count_t{2});
-                    return { pos.shift() - B, newn, ts, tail };
+                    return std::make_tuple( pos.shift() - B, newn, ts, tail );
                 } else {
                     if (mutate) {
                         pos.each_right(dec_visitor{}, idx + 1);
                         node->ensure_mutable_relaxed_n(e, idx)->d.count = idx;
-                        return { pos.shift(), node, ts, tail };
+                        return std::make_tuple( pos.shift(), node, ts, tail );
                     } else {
                         auto newn = node_t::copy_inner_r_e(e, node, idx);
                         if (Mutating) pos.visit(dec_visitor{});
-                        return { pos.shift(), newn, ts, tail };
+                        return std::make_tuple( pos.shift(), newn, ts, tail );
                     }
                 }
             } catch (...) {
@@ -962,29 +962,29 @@ struct slice_right_mut_visitor
                     if (mutate) {
                         node->inner()[idx] = next;
                         pos.each_right(dec_visitor{}, idx + 1);
-                        return { pos.shift(), node, ts, tail };
+                        return std::make_tuple( pos.shift(), node, ts, tail );
                     } else {
                         auto newn  = node_t::copy_inner_e(e, node, idx);
                         newn->inner()[idx] = next;
                         if (Mutating) pos.visit(dec_visitor{});
-                        return { pos.shift(), newn, ts, tail };
+                        return std::make_tuple( pos.shift(), newn, ts, tail );
                     }
                 } else if (idx == 0) {
                     if (Mutating) pos.visit(dec_right_visitor{}, count_t{1});
-                    return { pos.shift(), nullptr, ts, tail };
+                    return std::make_tuple( pos.shift(), nullptr, ts, tail );
                 } else if (Collapse && idx == 1 && pos.shift() > BL) {
                     auto newn = pos.node()->inner()[0];
                     if (!mutate) newn->inc();
                     if (Mutating) pos.visit(dec_right_visitor{}, count_t{2});
-                    return { pos.shift() - B, newn, ts, tail };
+                    return std::make_tuple( pos.shift() - B, newn, ts, tail );
                 } else {
                     if (mutate) {
                         pos.each_right(dec_visitor{}, idx + 1);
-                        return { pos.shift(), node, ts, tail };
+                        return std::make_tuple( pos.shift(), node, ts, tail );
                     } else {
                         auto newn = node_t::copy_inner_e(e, node, idx);
                         if (Mutating) pos.visit(dec_visitor{});
-                        return { pos.shift(), newn, ts, tail };
+                        return std::make_tuple( pos.shift(), newn, ts, tail );
                     }
                 }
             } catch (...) {
@@ -1007,15 +1007,15 @@ struct slice_right_mut_visitor
         auto mutate        = Mutating && node->can_mutate(e);
         if (new_tail_size == old_tail_size) {
             if (!Mutating) node->inc();
-            return { 0, nullptr, new_tail_size, node };
+            return std::make_tuple( 0, nullptr, new_tail_size, node );
         } else if (mutate) {
             destroy_n(node->leaf() + new_tail_size,
                       old_tail_size - new_tail_size);
-            return { 0, nullptr, new_tail_size, node };
+            return std::make_tuple( 0, nullptr, new_tail_size, node );
         } else {
             auto new_tail = node_t::copy_leaf_e(e, node, new_tail_size);
             if (Mutating) pos.visit(dec_visitor{});
-            return { 0, nullptr, new_tail_size, new_tail };
+            return std::make_tuple( 0, nullptr, new_tail_size, new_tail );
         }
     }
 };
@@ -1053,15 +1053,15 @@ struct slice_right_visitor : visitor_base<slice_right_visitor<NodeT, Collapse>>
                     newn->inner()[idx] = next;
                     newr->d.sizes[idx] = last + 1 - ts;
                     newr->d.count = count;
-                    return { pos.shift(), newn, ts, tail };
+                    return std::make_tuple( pos.shift(), newn, ts, tail );
                 } else if (idx == 0) {
-                    return { pos.shift(), nullptr, ts, tail };
+                    return std::make_tuple( pos.shift(), nullptr, ts, tail );
                 } else if (Collapse && idx == 1 && pos.shift() > BL) {
                     auto newn = pos.node()->inner()[0];
-                    return { pos.shift() - B, newn->inc(), ts, tail };
+                    return std::make_tuple( pos.shift() - B, newn->inc(), ts, tail );
                 } else {
                     auto newn = node_t::copy_inner_r(pos.node(), idx);
-                    return { pos.shift(), newn, ts, tail };
+                    return std::make_tuple( pos.shift(), newn, ts, tail );
                 }
             } catch (...) {
                 assert(!next || pos.shift() > BL);
@@ -1089,15 +1089,15 @@ struct slice_right_visitor : visitor_base<slice_right_visitor<NodeT, Collapse>>
                 if (next) {
                     auto newn  = node_t::copy_inner_n(idx + 1, pos.node(), idx);
                     newn->inner()[idx] = next;
-                    return { pos.shift(), newn, ts, tail };
+                    return std::make_tuple( pos.shift(), newn, ts, tail );
                 } else if (idx == 0) {
-                    return { pos.shift(), nullptr, ts, tail };
+                    return std::make_tuple( pos.shift(), nullptr, ts, tail );
                 } else if (Collapse && idx == 1 && pos.shift() > BL) {
                     auto newn = pos.node()->inner()[0];
-                    return { pos.shift() - B, newn->inc(), ts, tail };
+                    return std::make_tuple( pos.shift() - B, newn->inc(), ts, tail );
                 } else {
                     auto newn = node_t::copy_inner_n(idx, pos.node(), idx);
-                    return { pos.shift(), newn, ts, tail };
+                    return std::make_tuple( pos.shift(), newn, ts, tail );
                 }
             } catch (...) {
                 assert(!next || pos.shift() > BL);
@@ -1117,7 +1117,7 @@ struct slice_right_visitor : visitor_base<slice_right_visitor<NodeT, Collapse>>
         auto new_tail      = new_tail_size == old_tail_size
             ? pos.node()->inc()
             : node_t::copy_leaf(pos.node(), new_tail_size);
-        return { 0, nullptr, new_tail_size, new_tail };
+        return std::make_tuple( 0, nullptr, new_tail_size, new_tail );
     }
 };
 
@@ -1214,7 +1214,7 @@ struct slice_left_mut_visitor
                     node_t::inc_nodes(newn->inner() + 1, newcount - 1);
                     if (Mutating) pos.visit(dec_visitor{});
                 }
-                return { pos.shift(), newn };
+                return std::make_tuple( pos.shift(), newn );
             } catch (...) {
                 if (!mutate) node_t::delete_inner_r_e(newn);
                 throw;
@@ -1276,7 +1276,7 @@ struct slice_left_mut_visitor
                     node_t::inc_nodes(newn->inner() + 1, newcount - 1);
                     if (Mutating) pos.visit(dec_visitor{});
                 }
-                return { pos.shift(), newn };
+                return std::make_tuple( pos.shift(), newn );
             } catch (...) {
                 if (!mutate) node_t::delete_inner_r_e(newn);
                 else {
@@ -1305,11 +1305,11 @@ struct slice_left_mut_visitor
             auto newcount = count - idx;
             std::move(data + idx, data + count, data);
             destroy_n(data + newcount, idx);
-            return { 0, node };
+            return std::make_tuple( 0, node );
         } else {
             auto newn = node_t::copy_leaf_e(e, node, idx, count);
             if (Mutating) pos.visit(dec_visitor{});
-            return { 0, newn };
+            return std::make_tuple( 0, newn );
         }
     }
 };
@@ -1356,7 +1356,7 @@ struct slice_left_visitor : visitor_base<slice_left_visitor<NodeT, Collapse>>
                                         n->inner() + count,
                                         newn->inner() + 1);
                 node_t::inc_nodes(newn->inner() + 1, newr->d.count - 1);
-                return { pos.shift(), newn };
+                return std::make_tuple( pos.shift(), newn );
             } catch (...) {
                 node_t::delete_inner_r(newn, newc);
                 throw;
@@ -1368,7 +1368,7 @@ struct slice_left_visitor : visitor_base<slice_left_visitor<NodeT, Collapse>>
     static result_t visit_leaf(PosT&& pos, size_t first)
     {
         auto n = node_t::copy_leaf(pos.node(), pos.index(first), pos.count());
-        return { 0, n };
+        return std::make_tuple( 0, n );
     }
 };
 

--- a/immer/detail/rbts/rrbtree.hpp
+++ b/immer/detail/rbts/rrbtree.hpp
@@ -444,14 +444,14 @@ struct rrbtree
         using std::get;
         auto tail_off = tail_offset();
         if (idx >= tail_off) {
-            return { tail->leaf(), tail_off, size };
+            return std::make_tuple( tail->leaf(), tail_off, size );
         } else {
             auto subs = visit_maybe_relaxed_sub(
                 root, shift, tail_off,
                 region_for_visitor<T>(), idx);
             auto first = idx - get<1>(subs);
             auto end   = first + get<2>(subs);
-            return { get<0>(subs), first, end };
+            return std::make_tuple( get<0>(subs), first, end );
         }
     }
 

--- a/immer/detail/rbts/rrbtree.hpp
+++ b/immer/detail/rbts/rrbtree.hpp
@@ -302,7 +302,7 @@ struct rrbtree
             auto new_root = make_relaxed_pos(root, shift, r)
                 .visit(push_tail_visitor<node_t>{}, tail, tail_size);
             if (new_root)
-                return { shift, new_root };
+                return std::make_tuple( shift, new_root );
             else {
                 auto new_root = node_t::make_inner_r_n(2);
                 try {
@@ -316,7 +316,7 @@ struct rrbtree
                     node_t::delete_inner_r(new_root, 2);
                     throw;
                 }
-                return { shift + B, new_root };
+                return std::make_tuple( shift + B, new_root );
             }
         } else if (size == size_t{branches<B>} << shift) {
             auto new_root = node_t::make_inner_n(2);
@@ -328,13 +328,13 @@ struct rrbtree
                 node_t::delete_inner(new_root, 2);
                 throw;
             }
-            return { shift + B, new_root };
+            return std::make_tuple( shift + B, new_root );
         } else if (size) {
             auto new_root = make_regular_sub_pos(root, shift, size)
                 .visit(push_tail_visitor<node_t>{}, tail);
-            return { shift, new_root };
+            return std::make_tuple( shift, new_root );
         } else {
-            return { shift, node_t::make_path(shift, tail) };
+            return std::make_tuple( shift, node_t::make_path(shift, tail) );
         }
     }
 

